### ISSUE TITLE
What is the purpose of fp16 training? faster training? or better accuracy?

### ISSUE THREAD

**Issue by dedoogong** [pytorch/pytorch#25294]
I think for fp16 inference, just train a model on FP32 and then just model.half() will work.
But this forceful type casting would lead worse accuracy.

I'm not sure which is better between 
1. FP32 training then model.half() -> fp16 inference VS 
2. FP16 training using apex then inference FP16 with the same setting.

if 1 and 2 has not that big accuracy gap, case 1 would be used just for faster/memory efficent training?

Please give me any hint. 

Thank you.

**Comment by suvojit-0x55aa** [comment]
From some experiments I've run on MNIST with LeNet for 20 epochs, the 1st method will lead to some loss in accuracy of the model 
```
FP32: Acc: 96.790% (9679/10000)
Same weight converted to FP16: Acc: 96.690% (9669/10000)
Trained with FP16: Acc: 96.090% (9609/10000)
```
Although the accuracy gap is not much for dataset like MNIST using method (1) it should be more for real world dataset. Training on FP16 adapts the weight for FP16 inference so it should be able to match the accuracy with sufficient training.

Training on FP32 and converting it to FP16 will also throw off the weights as per my hunch. This can be checked by training using both methods and comparing the weight distribution.

Also training with FP16 on supported GPUs drastically cuts off training time while preserving accuracy as shown in [studies](https://devblogs.nvidia.com/tensor-cores-mixed-precision-scientific-computing/) and leads to smaller memory footprint.

I did a detailed comparison of the training on different GPUs, summarised in [suvojit-0x55aa/mixed-precision-pytorch](https://github.com/suvojit-0x55aa/mixed-precision-pytorch). Feel free to play around with the code and share your results.

**Comment by soumith** [comment]
please use https://discuss.pytorch.org for questions

**Comment by JensenHJS** [comment]
FP32 training then model.half() -> fp16 inference
This method indeed accelerate inference but it is far away from twice of fp32. My platform is nvidia tx2 and pytorch. Its compute capability is 6.2. I wonder why not. Thank you for your reply
